src/net/tcp/stream.rs

```diff
@@ -123,6 +124,40 @@ impl TcpStream {
         self.inner.nodelay()
     }
 
+    /// Sets the value for the `SO_LINGER` option on this socket.
+    ///
+    /// When set, the closing/shutting down of the socket will not return
+    /// until all queued messages for the socket have been successfully
+    /// sent or the linger timeout has been reached.
+    pub fn set_linger(&self, dur: Option<Duration>) -> io::Result<()> {
+        let sock = std::mem::ManuallyDrop::new(self.to_sock());
+
+        sock.set_linger(dur)
+    }
+
+    /// Gets the value of the `SO_LINGER` option on this socket.
+    ///
+    /// For more information about this option, see [`set_linger`][link].
+    ///
+    /// [link]: #method.set_linger
+    pub fn linger(&self) -> io::Result<Option<Duration>> {
+        let sock = std::mem::ManuallyDrop::new(self.to_sock());
+
+        sock.get_linger()
+    }
+
+    fn to_sock(&self) -> TcpSocket {
```

Comment:
Why not `impl From<TcpStream> for TcpSocket`? And `impl From<TcpSocket> for TcpStream`.

Reply:
sorry about that. I was just on autopilot as I basically had this change already in `tokio::net::TcpStream`.

I'll update this.